File: Lab 3/Java/SugestedProb15.java
import java.util.Scanner;

public class SugestedProb15 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt())
            return;
        int N = sc.nextInt();
        int A = sc.nextInt();
        int B = sc.nextInt();
        int C = sc.nextInt();

        long prev = A * 0 * 0 + B * 0 * 0 + C * 0 * 0;
        int current = 1;
        int ans = 1;

        for (int x = 0; x < N; x++) {
            for (int y = 0; y < N; y++) {
                if (x == 0 && y == 0)
                    continue;
                long val = (long) A * x * x + (long) B * y * y + (long) C * x * y;
                if (val > prev)
                    current++;
                else
                    current = 1;
                if (current > ans)
                    ans = current;
                prev = val;
            }
        }
        System.out.println(ans);
    }
}

File: Lab 3/SugestedProb15.py
import sys
input = sys.stdin.readline

N, A, B, C = map(int, input().split())

prev = A*0*0 + B*0*0 + C*0*0
current = 1
ans = 1

for x in range(N):
    for y in range(N):
        val = A*x*x + B*y*y + C*x*y
        if x == 0 and y == 0:
            continue
        if val > prev:
            current += 1
        else:
            current = 1
        ans = max(ans, current)
        prev = val

print(ans)